source/elements/oneMKL/source/domains/spblas/data_types/set_coo_matrix_data.rst

```diff
@@ -0,0 +1,209 @@
+``smhandle`` is destroyed with ``destroy_sparse_matrix`` or the data is reset with
+``set_coo_matrix_data``.
+
+In the case of USM, the object does not take ownership of the data.
+
+Also see :ref:`onemkl_sparse_create_coo_matrix`.
+
+.. _onemkl_sparse_set_coo_matrix_data_buffer:
+
+set_coo_matrix_data (Buffer version)
+------------------------------------
+
+.. rubric:: Syntax
+
+.. code-block:: cpp
+
+   namespace oneapi::mkl::sparse {
+
+       template <typename dataType, typename indexType>
+       void set_coo_matrix_data (sycl::queue                          &queue,
+                                 oneapi::mkl::sparse::matrix_handle_t smhandle,
+                                 std::int64_t                         num_rows,
+                                 std::int64_t                         num_cols,
+                                 std::int64_t                         nnz,
+                                 index_base                           index,
+                                 sycl::buffer<indexType, 1>           &row_ind,
+                                 sycl::buffer<indexType, 1>           &col_ind,
+                                 sycl::buffer<dataType, 1>            &val);
+
+   }
+
+.. container:: section
+
+   .. rubric:: Template parameters
+
+   dataType
+      See :ref:`supported template types<onemkl_sparse_supported_types>`. Can be
+      a different type than what was used when creating the ``matrix_handle_t``.
+
+   indexType
+      See :ref:`supported template types<onemkl_sparse_supported_types>`. Can be
+      a different type than what was used when creating the ``matrix_handle_t``.
+
+.. container:: section
+
+   .. rubric:: Input parameters
+
+   queue
+      The SYCL command queue which will be used for SYCL kernels execution.
+
+   smhandle
+      Handle already initialized with :ref:`onemkl_sparse_create_coo_matrix`.
+
+   num_rows
+      Number of rows of the provided data ``val``. Must be at least 0.
+
+   num_cols
+      Number of columns of the provided data ``val``. Must be at least 0.
+
+   nnz
+      Number of non-zero entries in the matrix (which may include explicit
+      zeros). Must be at least 0.
+
+   index
+      Indicates how input arrays are indexed. The possible options are described
+      in :ref:`onemkl_enum_index_base` enum class.
+
+   row_ind
+      Buffer of length ``nnz`` containing the row indices in ``index``-based
+      numbering. Refer to :ref:`onemkl_sparse_coo` format for detailed
+      description of ``row_ind``.
+
+   col_ind
+      Buffer of length ``nnz`` containing the column indices in ``index``-based
+      numbering. Refer to :ref:`onemkl_sparse_coo` format for detailed
+      description of ``col_ind``.
+
+   val
+      Buffer of length ``nnz`` containing non-zero elements (and possibly
+      explicit zeros) of the input matrix. Refer to :ref:`onemkl_sparse_coo`
+      format for detailed description of ``val``.
```

Comment:
Food for thought: Maybe this API can disallow 0 length/0-sized buffers, or nullptrs in the arguments. That is, calling this API should "complete" the sparse matrix handle. This API should also possibly be called once per matrix handle and no more than that, but I'll have to think about that.

Reply:
I would suggest that we let the freedom of resetting a handle to a nullptr state. I expect it can make things easier for users (in particular if we add getters in the future) and for the implementations.

I don't think of the `set_*_data` functions to be used only for completing a handle, it can also be used by users to cache handle structures and re-use them for different purposes. At least it would be nice to allow this IMO. Also that reminded me: we should add overloads for these functions that only reset the USM pointers or buffers. This would be more inlined with the cuSPARSE equivalent.

Reply:
I've not added the overloads that would only reset the pointers or buffers for now. We can consider adding them and some getters functions in a future version if needed.

Reply:
I'm keeping this unresolved because of your comment:

> Also that reminded me: we should add overloads for these functions that only reset the USM pointers or buffers. This would be more inlined with the cuSPARSE equivalent.

Reply:
In my latest comment I figured we can wait to add these functions if we have external feedback. I'm not sure how useful they are in practice. It looks like you agreed?

Reply:
Right, I'd forgotten I had hit the like button on this.